lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp

```diff
@@ -0,0 +1,466 @@
+    }
+  }
+
+  void fixUp(SILFunction *) {
+    for (auto &BB : getBuilder().getFunction()) {
+      for (auto &cloned : BB) {
+        // Fix up the type of try_apply successor block arguments.
+        if (auto *tryApply = dyn_cast<TryApplyInst>(&cloned)) {
+          auto normalBB = tryApply->getNormalBB();
+          SILFunctionConventions calleeConv(
+              tryApply->getSubstCalleeType(),
+              tryApply->getFunction()->getModule());
+          auto normalBBType = (*normalBB->args_begin())->getType();
+          auto applyResultType = calleeConv.getSILResultType();
+          if (normalBBType != calleeConv.getSILResultType()) {
+            auto origPhi = normalBB->getPhiArguments()[0];
+            SILValue undef =
+                SILUndef::get(normalBBType, getBuilder().getFunction());
+            SmallVector<Operand *, 8> useList(origPhi->use_begin(),
+                                              origPhi->use_end());
+            for (auto *use : useList) {
+              use->set(undef);
+            }
+
+            auto *newPhi = normalBB->replacePhiArgument(
+                0, applyResultType, origPhi->getOwnershipKind());
+
+            getBuilder().setInsertionPoint(normalBB->begin());
+            auto cast = createCast(tryApply->getLoc(), newPhi, normalBBType);
+            for (auto *use : useList) {
+              use->set(cast);
+            }
+          }
+        }
+      }
+    }
+  }
+};
+} // namespace
+
+void OpaqueSpecializerCloner::clone() {
+  for (auto arg: entryBlock->getArguments())
+    recordFoldedValue(arg, arg);
+  cloneReachableBlocks(cloneFromBlock, {}, entryBlock,
+                       true /*havePrepopulatedFunctionArgs*/);
+  getBuilder().setInsertionPoint(entryBlock);
+  getBuilder().createBranch(RegularLocation::getAutoGeneratedLocation(),
+                            getOpBasicBlock(cloneFromBlock));
+}
+
+/// Update address uses of the opaque type archetype with the concrete type.
+/// This is neccessary for apply instructions.
+void OpaqueSpecializerCloner::insertOpaqueToConcreteAddressCasts(
+    SILInstruction *orig, SILInstruction *cloned) {
+
+  // Replace apply operands.
+  if (auto apply = ApplySite::isa(cloned)) {
+    SavedInsertionPointRAII restore(getBuilder());
+    getBuilder().setInsertionPoint(apply.getInstruction());
+    auto substConv = apply.getSubstCalleeConv();
+    unsigned idx = 0;
+    for (auto &opd : apply.getArgumentOperands()) {
+      auto argConv = apply.getArgumentConvention(opd);
+      auto argIdx = apply.getCalleeArgIndex(opd);
+      auto argType = substConv.getSILArgumentType(argIdx);
+      if (argType.getASTType() != opd.get()->getType().getASTType()) {
+        if (argConv.isIndirectConvention()) {
+          auto cast = getBuilder().createUncheckedAddrCast(apply.getLoc(),
+                                                           opd.get(), argType);
+          opd.set(cast);
+        } else if (argType.is<SILFunctionType>()) {
```

Comment:
What if its an optional of a function type, or a tuple type, etc? You need to recursively transform types instead

Reply:
Good point! I will go through memory for now in those cases.

Reply:
https://github.com/apple/swift/pull/24436